ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java

```
@@ -0,0 +1,504 @@
+        for (Descriptors.FieldDescriptor fdef : this.descriptor.getFields()) {
+            if (fdef.isRepeated()) {
+                dup.addRepeatedField(fdef, this.getRepeatedField(context, fdef));
+            } else {
+                dup.builder.setField(fdef, this.builder.getField(fdef));
+
+            }
+        }
+        return dup;
+    }
+
+    /*
+     * call-seq:
+     *     Message.descriptor => descriptor
+     *
+     * Class method that returns the Descriptor instance corresponding to this
+     * message class's type.
+     */
+    @JRubyMethod(name = "descriptor", meta = true)
+    public static IRubyObject getDescriptor(ThreadContext context, IRubyObject recv) {
+        return ((RubyClass) recv).getInstanceVariable("@descriptor");
+    }
+
+    /*
+     * call-seq:
+     *     MessageClass.encode(msg) => bytes
+     *
+     * Encodes the given message object to its serialized form in protocol buffers
+     * wire format.
+     */
+    @JRubyMethod(meta = true)
+    public static IRubyObject encode(ThreadContext context, IRubyObject recv, IRubyObject message) {
+        RubyMessage rbVal = (RubyMessage) message;
+        return context.runtime.newString(new ByteList(rbVal.build(context).toByteArray()));
+    }
+
+    @JRubyMethod(meta = true)
+    public static IRubyObject decode(ThreadContext context, IRubyObject recv, IRubyObject data) {
+        byte[] bin = data.convertToString().getBytes();
+        RubyMessage ret = (RubyMessage) ((RubyClass) recv).newInstance(context, Block.NULL_BLOCK);
+        RubyDescriptor rubyDescriptor = (RubyDescriptor) ((RubyClass) recv).getInstanceVariable("@descriptor");
+        try {
+            DynamicMessage dynamicMessage = DynamicMessage.parseFrom(rubyDescriptor.getDescriptor(), bin);
+            ret.buildFrom(dynamicMessage);
+        } catch (InvalidProtocolBufferException e) {
+            e.printStackTrace(context.runtime.getOutputStream());
```

Instead of printing the stack trace to the console, can we throw a Ruby exception here?